components/support/base/src/main/java/mozilla/components/support/base/utils/LazyComponent.kt

```diff
@@ -0,0 +1,80 @@
+
+import androidx.annotation.VisibleForTesting
+import androidx.annotation.VisibleForTesting.PRIVATE
+import mozilla.components.support.base.log.logger.Logger
+import java.util.concurrent.atomic.AtomicInteger
+
+private val logger = Logger("LazyComponent")
+
+/**
+ * A wrapper around the [lazy] property delegate that is intended for use with application-wide
+ * components (such as services in the Service Locator pattern or dependencies in the Dependency
+ * Injection pattern).
+ *
+ * This class helps address the issue where kotlin's built-in [lazy] property delegate does not
+ * handle recursive initialization by inserting a getter between the lazy reference. For example, in:
+ * ```
+ * val useCases by lazy { UseCases(...) }
+ * val search by lazy { Search(useCases) }
+ * ```
+ * When `search` is used, it will initialized and, somewhat unintuitively, **so will `useCases`**.
+ * This is a problem if there are many dependencies or they are expensive. [LazyComponent]
+ * addresses this issue by allowing the class to be referenced without being initialized. To apply
+ * this to the example above:
+ * ```
+ * val useCases = LazyComponent { UseCases(...) }
+ * val search = LazyComponent { Search(useCases) }
+ * ```
+ *
+ * To call methods on the class and thus initialize it, one would call `LazyComponent.get()`, e.g.
+ * `search.get().startSearch(terms)`.
+ *
+ * This class also adds performance monitoring code to component initialization that, when paired
+ * with testing, can help prevent component initialization at inopportune times.
+ */
+class LazyComponent<T>(initializer: () -> T) {
+    // Lazy is thread safe.
+    private val lazyValue: Lazy<T> = lazy {
+        val initCount = initCount.incrementAndGet() // See property kdoc with regard to overflow.
+
+        initializer().also {
+            @Suppress("UNNECESSARY_NOT_NULL_ASSERTION") // the compiler fails with !! but warns with !!.
+            val className = if (it == null) "null" else it!!::class.java.canonicalName
+            logger.debug("Initialized lazyComponent #$initCount: $className")
+        }
+    }
+
+    /**
+     * Returns the component, initializing it if it has not already been initialized.
+     *
+     * To ensure this value is initialized lazily, it is expected that this method will only be
+     * called when the class instance needs to be interacted with, i.e. the return value should not
+     * be stored in a member property during an initializer.
+     */
+    fun get(): T = lazyValue.value
+
+    /**
+     * Returns whether or not the component has been initialized yet.
+     */
+    @VisibleForTesting(otherwise = PRIVATE)
+    fun isInitialized(): Boolean = lazyValue.isInitialized()
+
+    companion object {
+        /**
+         * The number of [LazyComponent]s initialized. This is intended to be checked during testing.
+         * For example, you can run a test scenario that starts the app and, if the [initCount]
```

Comment:
nit: would write this with the "you!", Java/KDoc should ideally be in third person...

Reply:
I changed it to "the team"